### DISCUSSION TITLE
Hexagonal architecture and Domain Driven Design

### DISCUSSION THREAD

**Discussion by LetoAtreides11235813** [techexcellenceio/techexcellence#10] [Show and tell]
Hello Valentina and the community,

As we already discussed with you Valentina earlier, I would like some feedback on this article I wrote on hex architecture and DDD:

https://towardsdev.com/hexagonal-architecture-and-domain-driven-design-bc2525dbc05f

It is to know if the reflection on Clean Architecture is still in the right direction and if the example is clear enough.

I will update the test case to fit what I saw in your banking project. I found your test architecture very useful

Thank you very much for your time and see you later

**Comment by irwansyahwii** [comment]
For DDD, AFAIK, you need to show the class diagram. The diagram should point out which ones are the aggregate roots, entities, and value object.

**Reply by LetoAtreides11235813** [reply]
Yes, thanks for this explanation.
I will grows the project for have more Domain classes and create diagram with all of theme

**Comment by valentinajemuovic** [comment]
@KevinDupeyrat thanks!

"When you translate DDD principels into Architecture Diagram, the most suitable approche is Microservices Architecture." 
--> Another equivalently valid option is a Modular Monolith.

Good that you have a GitHub example, I had a quick look:

1. I see NoArgs constructor for domain entities. Could you explain how you chose to do it this way, versus having a constructor?
2. How did you decide to put DateTimeUtil into the Domain, as opposte to infrastructure? (I see it as a presentation concern). Also why is releaseDate a string instead of being a date time type?
3. Regarding tests, I used to do it like this in the past, testing each class, but later I moved onto testing just use cases (in most situations).

**Comment by LetoAtreides11235813** [comment]
Hi @valentinacupac, thanks for your review.

1 - The NoArgsConstructor decorator is an error. i will delete it you are right
2 - For the utility class, I didn't know where to put it initially. For the moment, only the movie-infra module uses it but if it is needed in another module, where can I put it instead of Domain?
2 (BIS) - The releaseDate as a string comes directly from the API and I convert it to a date type when I receive it.
3 - Yes, I learned very recently that it is possible to create unit tests without a mirror source package. Indeed, in my experience, it was the only valid way to structure your test: One class = One test. I will use this project to learn how to create use case tests instead.

Thank you for your opinion and feel free to give me more in the future, I'm interested in learning more about clean coding and clean testing. That training my english too
